test/python_bindings.py

```diff
@@ -0,0 +1,435 @@
+
+        # Gruneisen parameters for copper
+        C0 = 0.394 * cm / us
+        S1 = 1.489
+        S2 = 0.
+        S3 = 0.
+        Gamma0 = 2.02
+        b = 0.47
+        rho0 = 8.93
+        T0 = 298.
+        P0 = 0.
+        Cv = 0.383e-05 * Mbcc_per_g
+
+        # Create the EOS
+        self.eos = singularity_eos.Gruneisen(C0, S1, S2, S3, Gamma0, b, rho0, T0, P0, Cv)
+
+        # Densities and energies
+        self.num = 4
+        self.density = np.zeros(self.num, dtype=np.double)
+        self.energy = np.zeros(self.num, dtype=np.double)
+
+        # Populate the input arrays
+        self.density[0] = 8.0
+        self.density[1] = 9.0
+        self.density[2] = 9.5
+        self.density[3] = 0.
+        self.energy[0] = 1.e9
+        self.energy[1] = 5.e8
+        self.energy[2] = 1.e8
+        self.energy[3] = 0.
+
+        # Gold standard values for a subset of lookups
+        self.pressure_true = np.array((-1.282094800000000e+11, 1.998504088912181e+10, 9.595823319513451e+10, P0 - C0 * C0 * rho0))
+        self.bulkmodulus_true = np.array((9.990648504000005e+11, 1.460692677162573e+12, 1.851227213843747e+12, Gamma0 * (P0 - C0 * C0 * rho0)))
+        self.temperature_true = np.array((5.590966057441253e+02, 4.285483028720627e+02, 3.241096605744125e+02, T0))
+        self.gruneisen_true = np.array((Gamma0, 2.007944444444444e+00, 1.927000000000000e+00, Gamma0))
+
+        self.temperature = np.zeros(self.num, dtype=np.double)
+        self.pressure = np.zeros(self.num, dtype=np.double)
+        self.bulkmodulus = np.zeros(self.num, dtype=np.double)
+        self.gruneisen = np.zeros(self.num, dtype=np.double)
+
+    def test_pressure(self):
+        """[VectorEOS][GruneisenEOS][Densities and Energies] A P(rho, e) lookup is performed"""
+        self.eos.PressureFromDensityInternalEnergy(self.density, self.energy, self.pressure, self.num)
+        assert_allclose(self.pressure, self.pressure_true, rtol=1e-12)
+
+    def test_bmod(self):
+        """[VectorEOS][GruneisenEOS][Densities and Energies] A B_S(rho, e) lookup is performed"""
+        self.eos.BulkModulusFromDensityInternalEnergy(self.density, self.energy, self.bulkmodulus, self.num)
+        assert_allclose(self.bulkmodulus, self.bulkmodulus_true, rtol=1e-12)
+
+    def test_temp(self):
+        """[Vector EOS][GruneisenEOS][Densities and Energies] A T(rho, e) lookup is performed"""
+        self.eos.TemperatureFromDensityInternalEnergy(self.density, self.energy, self.temperature, self.num)
+        assert_allclose(self.temperature, self.temperature_true, rtol=1e-12)
+
+    def test_gamma(self):
+        """[Vector EOS][GruneisenEOS][Densities and Energies] A Gamma(rho, e) lookup is performed"""
+        self.eos.GruneisenParamFromDensityInternalEnergy(self.density, self.energy, self.gruneisen, self.num)
+        assert_allclose(self.gruneisen, self.gruneisen_true, rtol=1e-12)
+
+
+class AluminumGruneisenEOS_SoundSpeedAndPressureComp(unittest.TestCase, EOSTestBase):
```

Comment:
The test coverage is great. I am concerned though that we basically have the same test code duplicated in python and C++.

I wonder if we could maintain the same level of coverage by removing some of this code in the C++ unit tests since the python version tests the same underlying code path.

What do you think? That can be a project for another time of course. We can merge this in first.

Reply:
One of the downsides of the Python variants is the missing GPU versions. Other than that, they're much easier to write. So probably having them in first won't hurt, but yes long term it would be nice to remove redundancies.